src/main/scala/mesosphere/marathon/core/storage/repository/BucketingRepository.scala

```diff
@@ -0,0 +1,162 @@
+package mesosphere.marathon
+package core.storage.repository
+
+import akka.NotUsed
+import akka.stream.scaladsl.{Flow, Source}
+import akka.util.ByteString
+import mesosphere.marathon.core.storage.zookeeper.PersistenceStore.Node
+import mesosphere.marathon.core.storage.zookeeper.ZooKeeperPersistenceStore
+
+import scala.concurrent.Future
+import scala.util.Try
+
+/**
+  * Fun fact about Zookeeper: you can create a lot of children nodes but if you  try to get all of them by calling
+  * [[mesosphere.marathon.core.storage.zookeeper.ZooKeeperPersistenceStore.children()]] (on the parent node) you are likely
+  * to get an error like:
+  * ```
+  * java.io.IOException: Packet len20800020 is out of range!
+  *  at org.apache.zookeeper.ClientCnxnSocket.readLength(ClientCnxnSocket.java:112)
+  *  ...
+  * ```
+  *
+  * Turns out that ZK has a packet length limit which is 4096 * 1024 bytes by default:
+  * https://github.com/apache/zookeeper/blob/0cb4011dac7ec28637426cafd98b4f8f299ef61d/src/java/main/org/apache/zookeeper/client/ZKClientConfig.java#L58
+  *
+  * It can be altered by setting `jute.maxbuffer` environment variable. Packet in this context is an application level
+  * packet containing all the children. Some experimentation showed that each child element in that packet has ~4bytes
+  * overhead for encoding. So, let's say each child node e.g. holding Marathon app definition is 50 characters long (
+  * seems like a good guess given 3-4 levels of nesting e.g. `eng_dev_databases_my-favourite-mysql-instance`), we could only
+  * have: 4096 * 1024 / (50 + 4) =  ~75k children nodes until we hit the exception.
+  *
+  * This trait solves the above mentioned limitations by putting the elements into buckets. Each element is then
+  * stored in `/$basePath/$bucket/$relative/element/path`. This way the total number of possible elements is
+  * NumberOfBuckets * MaxElementsPerBucket where later parameter depends on the length of children node name (see
+  * calculation above)
+  * Every element passed should implement [[BucketNode]] trait that defines the hash of the bucket where the element is stored
+  * and the relative path below.
+  *
+  * Same goes for reading elements - one need to pass a [[BucketPath]] which defines bucket hash along with the element
+  * relative path.
+  */
+trait BucketingRepository {
+
+  def store: ZooKeeperPersistenceStore
+
+  /**
+    * Max buckets number should never be changed in production since all data location will change and
+    * re-hashing will be needed
+    */
+  val MaxBuckets = 16
```

Comment:
How come 16 is enough? How many templates can we get with 16 buckets?

Reply:
See the [calculation](https://github.com/mesosphere/marathon/pull/6512/files#diff-78323e9f9f91a172f1d32a4c999554baR34) above - depending on the child's name length it's roughly 16 * 75k nodes.